net/src/ipc/n3h.rs

```diff
@@ -0,0 +1,294 @@
+        bail!(
+            "downloaded n3h version mismatch: {}, expected {}",
+            res,
+            &N3H_INFO.version
+        );
+    }
+
+    Ok(path)
+}
+
+/// check our pinned n3h version urls / hashes for the current os/arch
+fn get_artifact_info(os: &str, arch: &str) -> NetResult<&'static Artifact> {
+    let os = match os {
+        "linux" => &N3H_INFO.artifacts.linux,
+        "mac" => &N3H_INFO.artifacts.mac,
+        "win" => &N3H_INFO.artifacts.win,
+        _ => bail!("os {} not available", os),
+    };
+    if os.is_none() {
+        bail!("os not available");
+    }
+    let arch = match arch {
+        "aarch64" => &os.as_ref().unwrap().aarch64,
+        "x64" => &os.as_ref().unwrap().x64,
+        _ => bail!("arch {} not available", arch),
+    };
+    if arch.is_none() {
+        bail!("arch not available");
+    }
+    Ok(&arch.as_ref().unwrap())
+}
+
+/// run a command / capture the output
+fn exec_output<S1, I, S2>(cmd: S1, args: I, ignore_errors: bool) -> NetResult<String>
+where
+    S1: AsRef<std::ffi::OsStr>,
+    I: IntoIterator<Item = S2>,
+    S2: AsRef<std::ffi::OsStr>,
+{
+    let mut cmd = std::process::Command::new(cmd);
+    cmd.args(args);
+    tlog_d!("EXEC: {:?}", cmd);
+    let res = cmd.output()?;
+    if !ignore_errors && !res.status.success() {
+        bail!(
+            "bad exit {:?} {:?}",
+            res.status.code(),
+            String::from_utf8_lossy(&res.stderr)
+        );
+    }
+    Ok(String::from_utf8_lossy(&res.stdout).trim().to_string())
+}
+
+/// get the current os / arch
+fn get_os_arch() -> NetResult<(&'static str, &'static str)> {
+    if cfg!(target_os = "linux") && cfg!(target_arch = "x86_64") {
+        Ok(("linux", "x64"))
+    } else if cfg!(target_os = "linux") && cfg!(target_arch = "aarch64") {
+        Ok(("linux", "aarch64"))
+    } else if cfg!(target_os = "windows") && cfg!(target_arch = "x86_64") {
+        Ok(("win", "x64"))
+    } else if cfg!(target_os = "macos") && cfg!(target_arch = "x86_64") {
+        Ok(("mac", "x64"))
+    } else {
+        bail!("no prebuilt n3h for current os/arch - TODO check for node/npm - dl release zip");
+    }
+}
+
+/// extract a dmg archive
+fn extract_dmg(file: &std::ffi::OsStr, dest: &std::path::PathBuf) -> NetResult<std::path::PathBuf> {
+    let mut dest = dest.clone();
+    dest.push("n3h.app");
+    if !dest.exists() {
+        tlog_d!(
+            "{}",
+            exec_output(
+                "hdiutil",
+                vec![
+                    "attach",
+                    "-mountpoint",
+                    "./dmg-mount",
+                    &file.to_string_lossy()
+                ],
+                false
+            )?
+        );
+        exec_output(
+            "cp",
+            vec!["-a", "./dmg-mount/n3h.app", &dest.to_string_lossy()],
+            true,
+        )?;
+        tlog_d!(
+            "{}",
+            exec_output("hdiutil", vec!["detach", "./dmg-mount"], false)?
+        );
```

Comment:
I believe it would better to separate cmd execution from logging. e.g.
```Rust
    let output = exec_output("hdiutil", vec!["detach", "./dmg-mount"], false)?;
    tlog_d!("{}", output);
```
That way you are resilient to future changes of logging doctrine.

Reply:
:+1: